### DISCUSSION TITLE
Add Express example

### DISCUSSION THREAD

**Discussion by tomyo** [vercel/vercel#4594] [Ideas] [answered]
I want to make my team use vercel to deploy our new app, but they don't want to have to install vercel in their machine just to be able to run the local enviroment.
So, we want to build a express app, and run `yarn dev`. I want to know how to make the right vercel config file for this use case.

I know the new example section focused on zero-configuration.
Maybe then make a legacy examples repo, whatever.
But I believe that not including a express example somewhere, is a shoot on the foot.

**Comment by styfle** [comment, marked as answer]
Hi @tomyo 

We don't intend to add any examples for Express because this is an anti-pattern for the platform.

Serverless Functions already have [Node.js Helpers](https://vercel.com/docs/runtimes#official-runtimes/node-js/node-js-request-and-response-objects/node-js-helpers) which are similar to the Express API, no need to install any dependencies.

See this migration guide: https://vercel.com/guides/migrate-to-vercel

**Reply by lorenzofox3** [reply]
I am bit curious why you consider using frameworks such express, koa etc as an anti-pattern. If I am not wrong, the Node.js helpers do not provide middleware stack which can be useful for separation of concerns. 

```javascript
// api/my-handler.js
export default new koa()
   .use(loggerMiddleware)
   .use(someAuthMiddleware)
   .use(someDynamicCacheValueMiddleware)
   .use(actualHandler)
   .callback()
```

Moreover the ``headers`` we can provide with the route definition (in {now,vercel}.json) are static and therefore force us to handle some protocol related task at application level. Few simple cases: auth, dynamic cors (resolved from a DB), etc. I understand the "serverless" approach where one just has to focus on the business logic and the platform handles the rest, but in practice I struggle to figure out how to do so, even if I embrace the "cloud" way and delegate these tasks to thrid party services: at some point I still need to call them

**Reply by paulogdm** [reply]
@lorenzofox3 Full-blown APIs should use another provider, with a more suited offering around servers. Serverless functions should only be helpers and add-ons to your frontend, while complex APIs should be deployed elsewhere.

**Comment by tomyo** [comment]
Hi Steven, thanks.
There's one problem for us thought.
Vercel dev comand needs internet to run, and also, unlike Next which is at
least  a dependency in package.json, for zeit, an external app must be
downloaded and understood for the whole team.

We need a tool to easily deploy an existing express codebase. I understand
your point, but the team thinks rewriting the code just to make the new
cool deployment app happy is a no go.  I understand them too.
I know many months ago there was a way to manually tell now how to deploy
the files in the project, and that was it.

What can you advise me to do?

Thanks!

**Reply by revskill10** [reply]
You can use express as dev server, and use vercel to deploy, best of both worlds. It's what i'm doing for my projects.

**Comment by tomyo** [comment]
Hi Truong!
Can you provide an example?
The way code is written in both scenarios is quite different. i.e: routes.

Thanks!


On Tue, Jun 9, 2020 at 1:34 PM Truong Hoang Dung <notifications@github.com>
wrote:

> You can use express as dev server, and use vercel to deploy, best of both
> worlds. It's what i'm doing for my projects.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vercel/vercel/discussions/4594#discussioncomment-23948>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFM7WI2GD4LTNTBUNTGUPTRVYM5NANCNFSM4NYEEVNA>
> .
>


-- 
Tomyo

**Reply by revskill10** [reply]
OK, you will have lambda handlers. Then you require them in express server to be served in dev. And you will need lambda handlers to deploy to vercel, which requires the same lambda handlers.

**Comment by tomyo** [comment]
So, I would need a `routes.js` written express to replicate the urls from
`api/` folders in vercel, to the required lambda handler, and have a script
in package.json to use express with that file on development.
Didi I understand correctly?
Thanks.


On Tue, Jun 9, 2020 at 2:10 PM Truong Hoang Dung <notifications@github.com>
wrote:

> OK, you will have lambda handlers. Then you require them in express server
> to be served in dev. And you will need lambda handlers to deploy to vercel,
> which requires the same lambda handlers.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vercel/vercel/discussions/4594#discussioncomment-23953>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFM7WJFDO3JLAJSIDJEFZDRVYRB7ANCNFSM4NYEEVNA>
> .
>


-- 
Tomyo

**Reply by revskill10** [reply]
Yes, you're right. the only limitation is the api functions must be nodejs.
